src/ZODB/Connection.py

```diff
@@ -101,11 +103,10 @@ def __init__(self, db, cache_size=400, before=None, cache_size_bytes=0):
         # Multi-database support
         self.connections = {self._db.database_name: self}
 
-        storage = db._mvcc_storage
         if before:
-            storage = storage.before_instance(before)
+            storage = HistoricalStorageAdapter(db.storage, before)
```

Comment:
I'm slightly concerned about the asymmetry between the historical case and the normal case. I think I've convinced myself that the MVCC stuff (and keeping track of the instances in use) doesn't matter for the before connections because they're ReadOnly. Would this work instead as a simplification?

```python
storage = db._mvcc_storage.new_instance() # always get new instance
if before:
   storage = HistoricalStorageAdapter(storage, before)
```